ERC20ForSPL/contracts/ERC20ForSPLMintableFactory.sol

```diff
@@ -0,0 +1,108 @@
+/// @dev This contract is built with forked OpenZeppelin's UUPS standard and it's a Beacon contract at the same time. The storage is defined in the following way:
+/// @dev Storage slot 0 - taken by the forked BeaconProxy's implementation.
+/// @dev Storage slot 1 - taken by the forked UUPS's implementation.
+/// @dev Storage slot 2 - taken by the forked OwnableUpgradeable's owner.
+/// @dev Every next slot is defined by the needs of the ERC20ForSPLMintableFactory.
+/// @custom:oz-upgrades-unsafe-allow constructor
+contract ERC20ForSPLMintableFactory is OwnableUpgradeable, UUPSUpgradeable {
+    address private _implementation;
+    address private _uupsImplementation;
+    address private _owner;
+    mapping(bytes32 => address) public tokensData;
+    address[] public tokens;
+
+    event TokenDeploy(bytes32 indexed tokenMint, address indexed token);
+    event Upgraded(address indexed implementation);
+
+    error BeaconInvalidImplementation(address implementation);
+
+    /// @notice Disabling the initializers to prevent the implementation getting hijacked
+    constructor() {
+        _disableInitializers();
+    }
+
+    /// @notice Method used by the OpenZeppelin's BeaconProxy lib that mimics the functionality of a constructor
+    /// @param implementation_ The address of the BeaconProxy initial implementation
+    function initialize(address implementation_) public initializer {       
+        __Ownable_init(msg.sender);
+         _setImplementation(implementation_);
+    }
+
+    function _authorizeUpgrade(address) internal override onlyOwner {}
+
+    /**
+     * @dev Returns the current implementation address.
+     */
+    function implementation() public view virtual returns (address) {
+        return _implementation;
+    }
+
+    /**
+     * @dev Upgrades the beacon to a new implementation.
+     *
+     * Emits an {Upgraded} event.
+     *
+     * Requirements:
+     *
+     * - msg.sender must be the owner of the contract.
+     * - `newImplementation` must be a contract.
+     */
+    function upgradeTo(address newImplementation) public virtual onlyOwner {
+        _setImplementation(newImplementation);
+    }
+
+    /**
+     * @dev Sets the implementation contract address for this beacon
+     *
+     * Requirements:
+     *
+     * - `newImplementation` must be a contract.
+     */
+    function _setImplementation(address newImplementation) private {
+        if (newImplementation.code.length == 0) {
+            revert BeaconInvalidImplementation(newImplementation);
+        }
+        _implementation = newImplementation;
+        emit Upgraded(newImplementation);
+    }
+
+    /// @notice Deploys a new ERC20ForSPLMintable's BeaconProxy instance
+    /// @param _name The name of the SPLToken
+    /// @param _symbol The symbol of the SPLToken
+    /// @param _decimals The decimals of the SPLToken. This value cannot be bigger than 9, because of Solana's maximum value limit of uint64
```

Comment:
Why decimals can't be more 9? 
Look https://solscan.io/token/7aESLpU3s9AmUMobLqEHcdHPtwB6q5D32gznRGDnrHgV?cluster=devnet
And in our tests we used decimals 18 too for testing

Reply:
@gigimon because of the this **[docs page](https://docs.neonevm.org/docs/developing/deploy_facilities/interacting_with_spl_tokens#restrictions)**? tbh I've never tested erc20forspl with more than 9 decimals

Reply:
@anton-lisanin what do you think about this?